extensions/message-tags.md

```diff
@@ -54,6 +54,8 @@ Individual tag keys MUST only be used a maximum of once per message. Clients
 receiving messages with more than one occurrence of a tag key SHOULD discard all
 but the final occurrence.
 
+A tag with an empty value (e.g. `foo=`) MAY be considered equivalent to a tag missing a value (e.g. `foo`). Specifications MUST NOT assign different meanings to tags with empty or missing values.
```

Comment:
I think this should be a `MUST` and (to make the message smaller) the `foo` form should be stated as the canonical form.

Reply:
Imagine a basic parser looking like this:

```python
tags = 'foo;bar='
[part.split('=') for part in tags.split(';')]
# [['foo'],['bar','']]
```

Saying that implementations MAY consider empty/missing as equivalent *allows* a normalisation step without *requiring* it, and saying that specs MUST NOT differentiate makes that safe.

Changing the MAY to a MUST seems like forcing extra awkwardness on implementations for no real gain. Implementations might decide there's a gain *for them* but the spec certainly doesn't need it.

Reply:
It's not incompatible with a MUST to parse it like example but then check the values properly if they get compared.

Reply:
As currently written it feels like it's saying that specs can't assign different meanings to a tag when it has a value vs the same tag when it's not given a value at all, when what's actually intended is that specs can't treat the `a=` and `a` case differently. Maybe we could use the term _"tags with empty **vs** missing values"_ or similar to make it clear what we mean here?

For what it's worth I don't mind making it a MUST to ensure they're treated the same way, but I'm unsure about specifying which one software has to send out.

Reply:
There is a potential issue with a client tag `+a` being rewritten by a server to `+a=` and then the whole set of client tags may not fit in the outbound buffer. We should state that servers MUST NOT rewrite relayed client tags with a missing value from `+a` to `+a=`. This also applies more generally (i.e. not adding spurious backslashes).

Reply:
The size limits already forbid this. Let’s not bog down specs by repeating things covered elsewhere ad nauseam to cover minor edge cases with non compliant implementations.

PS also this thread applies to outdated code so is hidden, worth replying elsewhere.

Reply:
"The server MAY unescape tag values after receiving data and escape those values before sending them out."

> The size limits already forbid this.

Where is this specified?

> non compliant implementations

It's not non-compliant; `+a` is equivalent to `+a=` so a server implementation that finds it easier to always output `+a=` can do so.

(These comments aren't hidden on the PR page.)

Reply:
Read the size limit section. Servers and clients are forbidden from overflowing the output buffers. Hence, non-compliant.

Reply:
I'm not suggesting the server would overflow it, but could be forced to omit a client tag when it doesn't fit on the end because of missing values being rewritten as empty values.

Reply:
I don't think this possibility is worth us worrying about at all in the specification.

Reply:
Should be resolved by https://github.com/ircv3/ircv3-specifications/pull/373/commits/ed7ad5bb33f835708f3383d52c01643c4e4e1898